aya/src/symbols.rs

```diff
@@ -0,0 +1,688 @@
+            process_id: symbol.process_id(),
+        };
+
+        let mut guard = self.cache.borrow_mut();
+        match guard.get_mut(&key) {
+            Some(result) => {
+                *symbol = result.clone();
+            }
+            None => {
+                self.inner.resolve(symbol);
+                // TODO: Since we store potentially incomplete SymbolInfo here,
+                // we might miss out the chance to get later on additional information
+                guard.insert(key, symbol.clone());
+            }
+        }
+    }
+}
+
+/// The error type that is used when resolving symbols using the addr2line library
+#[derive(Debug, Error)]
+pub enum Addr2LineError {
+    /// Failed to read the context of the executable or library
+    #[error("Failed to read object data at path {path}")]
+    InvalidObjectPath {
+        /// The object path
+        path: PathBuf,
+        /// The original io::Error
+        source: io::Error,
+    },
+    /// Failed to parse data for an executable or library
+    #[error("Failed to read object data at path {path}")]
+    ParseError {
+        /// The object path
+        path: PathBuf,
+        /// The original error
+        source: addr2line::object::Error,
+    },
+    /// Failed to create the gimli resolver context
+    #[error("Failed to read object data at path {path}")]
+    CreateContextError {
+        /// The object path
+        path: PathBuf,
+        /// The original error
+        source: addr2line::gimli::Error,
+    },
+}
+
+struct Addr2LineObjectContext {
+    ctx: Context<EndianRcSlice<RunTimeEndian>>,
+}
+
+impl Addr2LineObjectContext {
+    pub fn from_object_path(path: &Path) -> Result<Self, Addr2LineError> {
+        let data = std::fs::read(path).map_err(|e| Addr2LineError::InvalidObjectPath {
+            path: path.to_owned(),
+            source: e,
+        })?;
+        let object: addr2line::object::File<_> = addr2line::object::File::parse(&data[..])
+            .map_err(|e| Addr2LineError::ParseError {
+                path: path.to_owned(),
+                source: e,
+            })?;
+        let ctx = Context::new(&object).map_err(|e| Addr2LineError::CreateContextError {
+            path: path.to_owned(),
+            source: e,
+        })?;
+
+        Ok(Self { ctx })
+    }
+}
+
+impl Addr2LineObjectContext {
+    pub fn resolve(&self, address: u64) -> Option<String> {
+        match self.ctx.find_frames(address) {
+            Ok(mut frames) => {
+                let mut result = None;
+
+                while let Ok(Some(frame)) = frames.next() {
+                    let frame = frame.function.and_then(|function_name| {
+                        function_name
+                            .demangle()
+                            .map(|demangled_name| demangled_name.to_string())
+                            .ok()
+                    });
+
+                    // Return the last frame in the stack of frames, since the inlined functions
```

Comment:
Now this part is an interesting design choice: Should the aya resolver return just the outermost frame? Or potentially all the frames? The latter might provide a bit more data for profilers that run on top of that, or for debugging purposes. But there's also more data produced - and the `SymbolResolver` API would need to be changed to be able to return a stack of symbols. Not exactly sure how that API should look like, since just returning `Vec<SymbolInfo>` might then have wrongly populated virtual  addresses if we just copy&paste those. Maybe the API would need to be closer to `resolve(&SymbolInfo) -> Vec<ResolvedSymbol>` where the latter doesn't care about PIDs and virtual addresses anymore?

Besides that producing more data obviously also has a cost.